[ty] Recognize overlapping NewTypes and their underlying values (#27522)
astral-sh/ruff: crates/ty_python_semantic/resources/mdtest/type_properties/is_disjoint_from.md

static_assert(not is_disjoint_from(str, LiteralString))
```

## NewTypes and literal types

A `NewType` overlaps with any literal compatible with its concrete base. Type checkers accept
`UserId(True)` and cannot reliably prevent such calls because `bool` is a subtype of `int`, so an

> integer-based `NewType` also overlaps boolean literals.

Reviewer: again, this makes it sound like the _ideal_ situation would be for type checkers to reject things like `UserId(True)`, but I don't understand why that would be the case; this would be a violation of Liskov and inconsistent with how type checkers work in all other situations

```py
from typing import Literal, NewType
from ty_extensions import static_assert
from ty_extensions._internal import is_disjoint_from

UserId = NewType("UserId", int)
StringId = NewType("StringId", str)
BytesId = NewType("BytesId", bytes)

UserId(True)

static_assert(not is_disjoint_from(UserId, Literal[True]))
static_assert(not is_disjoint_from(Literal[True], UserId))
static_assert(not is_disjoint_from(UserId, Literal[False]))
static_assert(not is_disjoint_from(Literal[False], UserId))
static_assert(not is_disjoint_from(UserId, Literal[True, False]))
static_assert(not is_disjoint_from(UserId, Literal[1]))
static_assert(not is_disjoint_from(Literal[1], UserId))
static_assert(not is_disjoint_from(UserId, bool))
static_assert(not is_disjoint_from(bool, UserId))
static_assert(not is_disjoint_from(UserId, int))
static_assert(is_disjoint_from(UserId, Literal["user"]))
static_assert(is_disjoint_from(Literal["user"], UserId))

static_assert(not is_disjoint_from(StringId, Literal["user"]))
static_assert(not is_disjoint_from(BytesId, Literal[b"user"]))
```

Nested NewTypes retain the overlap, and a float-based NewType also accepts `int` and `bool` through

> the numeric tower.

Reviewer: ```suggestion
the `int`/`float` special case.
```

there is no typing numeric tower; PEP 484 specifically rejects the numeric tower of PEP 3141 and instead proposes a "simple special case" instead of creating its own numeric tower

```py
NestedUserId = NewType("NestedUserId", UserId)
FloatId = NewType("FloatId", float)
BoolId = NewType("BoolId", bool)

static_assert(not is_disjoint_from(NestedUserId, bool))
static_assert(not is_disjoint_from(NestedUserId, Literal[True]))
static_assert(not is_disjoint_from(FloatId, bool))
static_assert(not is_disjoint_from(FloatId, Literal[True]))
static_assert(not is_disjoint_from(FloatId, Literal[1]))
static_assert(not is_disjoint_from(FloatId, int))
static_assert(not is_disjoint_from(FloatId, float))
static_assert(not is_disjoint_from(BoolId, bool))
```

## Distinct NewTypes

Distinct `NewType`s are not assignable to each other, but they overlap whenever their concrete bases

> overlap. They are disjoint only when their concrete bases are disjoint.

Reviewer: this part I do still question a little, since the _only_ way of getting a variable to be considered to inhabit two `NewType`s simultaneously when one is not a sub-NewType of the other is to use a `TypeIs` function, and ideally I think we would complain about such a `TypeIs` function as being invalidly defined. So it does seem to me that such a pair of NewTypes really is conceptually disjoint.

It's sort of interesting that the `NewType` constructor itself acts like a `TypeGuard` function rather than a `TypeIs` function. All type checkers agree that the type of `y` is `N2`, not `N1 & N2`:

```py
from typing import NewType

N1 = NewType("N1", int)
N2 = NewType("N2", int)

def f(x: N1):
    y = N2(x)
    reveal_type(y)  # N2, *not* N1 & N2
```

the `y = N2(x)` call is sort-of equivalent to a call to this function:

```py
def __create_N2(x: int) -> TypeGuard[N2]: ...
```

Author: I find it hard to justify doing the rest of this PR _without_ this part, though. If we say that `UserId` (a branded `int`) is not disjoint from `bool`, and we say that `MediaId` (another branded `int`) is also not disjoint from `bool`, then how can we claim `UserId` and `MediaId` are disjoint types?

Reviewer: Well, one distinction is that we provide sound ways to narrow a variable such that it is simultaneously considered to inhabit `bool` and `UserId`, but we provide no such way to soundly narrow a variable so that it can be simultaneously considered to inhabit both `UserId` and `MediaId`.

And another distinction is that NewType subtyping is of course nominal, but it's not subclassing-based, it's fundamentally different to runtime-subclassing-based subtyping. Since subtyping between a pair of NewTypes works fundamentally differently to how it works between a NewType and a regular nominal-instance, it doesn't seem inconsistent to me to say that disjointness between a pair of NewTypes also works fundamentally differently to how it works between a NewType and a regular nominal-instance.

To me it seems quite intuitive to say that two brands are mutually exclusive unless one explicitly subtypes the other, even if any one brand might overlap with nominal-instance subclasses of their concrete base.

And it's paralleled by this, I think? Some subtypes of `Base` are mutually disjoint but others are not in the following example:

```py
from typing import Literal, final, Self
from ty_extensions._internal import is_disjoint_from
from ty_extensions import static_assert

class Base: ...
class Sub1(Base): ...
class Sub2(Base): ...
@final
class Sub3(Base): ...

static_assert(not is_disjoint_from(Sub1, Sub2))
static_assert(is_disjoint_from(Sub1, Sub3))
```

And I think we could say in just the same way that `UserId` is not disjoint from `bool` but is disjoint from `MediaId`

Reviewer: Anyway, I don't think this should block the PR if you still disagree, it seems very unlikely to ever really matter much!

Author: I think you are correct here; I was wrong. Unless we allow a single value to be two different NewType at once (which I think we should not) then we should consider those distinct NewType to be disjoint. This makes the "tags" model @jelle-openai discusses below fully consistent, I think. I'll say more about that in a separate comment below.

```py
from typing import NewType
from ty_extensions import static_assert
from ty_extensions._internal import is_assignable_to, is_disjoint_from, is_subtype_of

First = NewType("First", int)
Second = NewType("Second", int)
NestedFirst = NewType("NestedFirst", First)
Numeric = NewType("Numeric", float)
Text = NewType("Text", str)

static_assert(not is_disjoint_from(First, Second))
static_assert(not is_disjoint_from(Second, First))
static_assert(not is_disjoint_from(NestedFirst, Second))
static_assert(not is_disjoint_from(First, Numeric))
static_assert(is_disjoint_from(First, Text))
static_assert(is_disjoint_from(Text, First))

static_assert(not is_subtype_of(First, Second))
static_assert(not is_subtype_of(Second, First))
static_assert(not is_assignable_to(First, Second))
static_assert(not is_assignable_to(Second, First))
```

## NewTypes and nominal subclasses

A `NewType` overlaps with subclasses of its base, whether or not those subclasses are final. It also

> overlaps with its own base and with any supertypes of that base.

Reviewer: codex just has finality in its head because of earlier iterations of this PR, but first-time readers of this document won't understand why we're talking about final classes here

```suggestion
A `NewType` overlaps with subclasses of its base. It also
overlaps with its own base and with any supertypes of that base.
```

static_assert(not is_disjoint_from(UserId, int))
static_assert(not is_disjoint_from(UserId, object))
static_assert(is_disjoint_from(UserId, str))
static_assert(not is_disjoint_from(FinalIntId, FinalInt))
static_assert(not is_disjoint_from(FinalIntId, int))
```

An `IntEnum` and its members also overlap with an integer-based `NewType`.

```py
from enum import IntEnum

class Choice(IntEnum):
    FIRST = 1
    SECOND = 2

static_assert(not is_disjoint_from(UserId, Choice))
static_assert(not is_disjoint_from(UserId, Literal[Choice.FIRST]))
static_assert(not is_disjoint_from(Literal[Choice.FIRST], UserId))
```

## NewTypes and generic classes

```toml
[environment]
python-version = "3.12"
```

A `NewType` based on a covariant generic specialization overlaps with its generic supertypes and
subclasses. Two differently specialized covariant types can also overlap through a common, more
specific specialization.

```py
from typing import Any, NewType, final
from ty_extensions import static_assert
from ty_extensions._internal import is_disjoint_from, is_subtype_of

class Base[T]:
    def get(self) -> T:
        raise NotImplementedError

> @final

Reviewer: why is this decorated with `@final`? is that relevant to the test at all?

class FinalChild[T](Base[T]): ...

class OrdinaryChild[T](Base[T]): ...

BaseId = NewType("BaseId", Base[int])

static_assert(not is_disjoint_from(BaseId, Base[int]))
static_assert(not is_disjoint_from(BaseId, Base[object]))

> static_assert(not is_disjoint_from(BaseId, Base[str]))

Reviewer: might be worth a comment here that this is because `Base[Never]` is an inhabited type, so `Base[str] & Base[int]` is inhabited even though `str & int` is not

# `A | Any` cannot materialize to be equivalent to `B`.
static_assert(is_disjoint_from(Invariant[A | Any], Invariant[B]))
static_assert(is_disjoint_from(Invariant[B], Invariant[A | Any]))
static_assert(is_disjoint_from(Invariant[A & Any], Invariant[B]))
static_assert(is_disjoint_from(Invariant[B], Invariant[A & Any]))
static_assert(is_disjoint_from(InvariantPair[A, A], InvariantPair[A, B]))
static_assert(not is_disjoint_from(Covariant[A], Covariant[B]))
static_assert(not is_disjoint_from(Covariant[A], CoSubB))
static_assert(not is_disjoint_from(Sequence[int], Sequence[str]))
```

### Specialized `@final` types

Final generic specializations can overlap through a shared subtype such as `Foo[Never]`.

```toml
[environment]
python-version = "3.12"
```

```py
from typing import Any, final
from ty_extensions import static_assert
from ty_extensions._internal import is_disjoint_from

@final
class Foo[T]:
    def get(self) -> T:
        raise NotImplementedError

class A: ...
class B: ...

static_assert(not is_disjoint_from(A, B))
static_assert(not is_disjoint_from(Foo[A], Foo[B]))
static_assert(not is_disjoint_from(Foo[A], Foo[Any]))
static_assert(not is_disjoint_from(Foo[Any], Foo[B]))

> # `Foo[Never]` is a subtype of both `Foo[int]` and `Foo[str]`.

Reviewer: ```suggestion
# `Foo[Never]` is a (possibly inhabited) subtype of both `Foo[int]` and `Foo[str]`.
```

static_assert(not is_disjoint_from(IntListId, list[Any]))
static_assert(is_disjoint_from(IntListId, list[str]))
static_assert(is_disjoint_from(IntListId, AnyListId))
```

A generic type variable must not make a potentially compatible specialization appear disjoint.
Compatible constraints and bounds also preserve the overlap.

```py
def unconstrained[T]() -> None:
    static_assert(not is_disjoint_from(IntListId, list[T]))
    static_assert(not is_disjoint_from(list[T], IntListId))

def compatible_constraints[T: (int, str)]() -> None:
    static_assert(not is_disjoint_from(IntListId, list[T]))

def compatible_bound[T: int]() -> None:
    static_assert(not is_disjoint_from(IntListId, list[T]))
```

## Generic aliases

Generic class objects and aliases overlap compatible `type[...]` types; incompatible invariant
specializations make them disjoint.

```toml
[environment]
python-version = "3.12"
```

```py
from typing import final
from ty_extensions import static_assert
from ty_extensions._internal import TypeOf, is_disjoint_from

class GenericClass[T]:
    x: T  # invariant

static_assert(not is_disjoint_from(TypeOf[GenericClass], type[GenericClass]))  # error: [missing-type-argument]

> static_assert(not is_disjoint_from(TypeOf[GenericClass[int]], type[GenericClass]))  # error: [missing-type-argument]

Reviewer: Or just add the missing type argument to avoid the unrelated error? (And similar elsewhere in this snippet)

```suggestion
static_assert(not is_disjoint_from(TypeOf[GenericClass[int]], type[GenericClass[Any]]))
```
